lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -160,8 +179,8 @@ def convert_json(value: dict) -> dict:
 
 def convert_string(value: str) -> str:
     """
-    convert_string Verify that the parameter value is a string (e.g.,
-    not a JSON dict, or an int), and return it.
+    Verify that the parameter value is a string (e.g., not a JSON dict, or an
+    int), and return it.
```

Comment:
Yes, I slide between syntax and semantics. We require JSON data. The Python standard representation of JSON data is a `dict`; but a "JSON dict" is constrained as it requires string keys while values are either scalar primitives, an array, or another JSON dict. What we really need is a "JSON" typedef, something like

```
from typing import Dict, List, Union
JSONVALUE = Union[int,str,List["JSONVALUE"],"JSON"]
JSONKEY = str
JSON = Dict[JSONKEY,JSONVALUE]

x : JSON = {"a": {"b": False, "c": 1}, "d": [1, 2, 3]}
```

Maybe I'll try it that way ...

Reply:
I support that approach.  However, I would offer a couple of tweaks:
```
from typing import Dict, List, Union
JSONSTRING = str
JSONNUMBER = Union[int, float]
JSONVALUE = Union["JSONOBJECT", "JSONARRAY", JSONSTRING, JSONNUMBER, bool, None]
JSONARRAY = List[JSONVALUE]
JSONOBJECT = Dict[JSONSTRING,JSONVALUE]
JSON = JSONVALUE

x : JSON = {"a": {"b": False, "c": 1}, "d": [1, 2, 3], "e": 0.5, "f": False, "g": True, "h": None}
```

Reply:
I believe I've painted myself into a corner here for publish. I'm going to pretend the corner isn't there, and finesse this by returning `{"dataset": Dataset}`, close my eyes, and try really hard to pretend that's JSON, but it's not. In theory I could make it JSON by serializing the `Dataset`, and deserializing it later, but that seems ... excessive. Hmm.

Reply:
Would it make sense to use Python datatypes (either `dict` or our new `"JSON"` type) in the code and then transform it to "JSON" (whatever that means) just before handing it off to ES?  (That is, if ES wants the output of `json.dumps()`, we can insert that at the last step.)

I think manipulating JSON as a `dict` is much easier than doing operations on the stringized version of it.

Reply:
Yes, I agree that serializing and deserializing `Dataset` just so that we can tell ourselves that the container is a JSON object is too much.  I have no problem with us claiming, within our own code, that a `JSONOBJECT` is a container that we can stick anything we want into; we only need to be orthodox when we pass the object on to something like ES.

Reply:
Or, maybe we should just call it a `dict` instead of a `JSONOBJECT`...and then there should be no issue, right?

Reply:
For a moment there, I thought I could get around it by just using the JSON-ified document map, since `assemble` and `postprocess` don't use anything else.

Then I realized that was a bug: I'd meant to have `postprocess` actually update the `Dataset` on success. Oops. But that means I also need it to know the new access keyword, so I added a second key to the `context` `JSON`.

And, yes, I could make `context` a simple (what, `Dict[str,Any]`?), which would avoid that technicality, and have the "advantage" of going back to invalidating the `True` default that you dislike. ;-) I suppose that would be cleaner than going to all the trouble to create the `JSON` typedef and then fudging the use of it. (In theory, a static type checker, if we ever start using one, would flag my hack.)